bigchaindb/backend/localmongodb/connection.py

```diff
@@ -1,5 +1,268 @@
+                                       self.certfile,
+                                       self.keyfile,
+                                       self.keyfile_passphrase,
+                                       self.crlfile)
+
+            # FYI: the connection process might raise a
+            # `ServerSelectionTimeoutError`, that is a subclass of
+            # `ConnectionFailure`.
+            # The presence of ca_cert, certfile, keyfile, crlfile implies the
+            # use of certificates for TLS connectivity.
+            if self.ca_cert is None or self.certfile is None or \
+                    self.keyfile is None or self.crlfile is None:
+                client = pymongo.MongoClient(self.host,
+                                             self.port,
+                                             replicaset=self.replicaset,
+                                             serverselectiontimeoutms=self.connection_timeout,
+                                             ssl=self.ssl,
+                                             **MONGO_OPTS)
+                if self.login is not None and self.password is not None:
+                    client[self.dbname].authenticate(self.login, self.password)
+            else:
+                logger.info('Connecting to MongoDB over TLS/SSL...')
+                client = pymongo.MongoClient(self.host,
+                                             self.port,
+                                             replicaset=self.replicaset,
+                                             serverselectiontimeoutms=self.connection_timeout,
+                                             ssl=self.ssl,
+                                             ssl_ca_certs=self.ca_cert,
+                                             ssl_certfile=self.certfile,
+                                             ssl_keyfile=self.keyfile,
+                                             ssl_pem_passphrase=self.keyfile_passphrase,
+                                             ssl_crlfile=self.crlfile,
+                                             ssl_cert_reqs=CERT_REQUIRED,
+                                             **MONGO_OPTS)
+                if self.login is not None:
+                    client[self.dbname].authenticate(self.login,
+                                                     mechanism='MONGODB-X509')
+
+            return client
+
+        # `initialize_replica_set` might raise `ConnectionFailure`,
+        # `OperationFailure` or `ConfigurationError`.
+        except (pymongo.errors.ConnectionFailure,
+                pymongo.errors.OperationFailure) as exc:
+            logger.info('Exception in _connect(): {}'.format(exc))
+            raise ConnectionError(str(exc)) from exc
+        except pymongo.errors.ConfigurationError as exc:
+            raise ConfigurationError from exc
+
+
+MONGO_OPTS = {
+    'socketTimeoutMS': 20000,
+}
+
+
+def initialize_replica_set(host, port, connection_timeout, dbname, ssl, login,
```

Comment:
@vrde IIRC there we disscussions around removing replica set? If yes, then should we do that in this PR?